race-buffer/src/async_reader.rs

```diff
@@ -0,0 +1,286 @@
+
+    /// Take a snapshot of the RaceBuffer's backing storage at the given index
+    fn snap_storage(&self, index: usize) -> Result<E, Self::Error>;
+}
+
+/// Used to read from a RaceBuffer asynchronously
+pub struct RaceReader<E, S>
+where
+    E: Entry,
+    S: Snapper<E>,
+{
+    /// Struct for reading writer state
+    snapper: S,
+    /// Sequence number of next entry to be read
+    read_seqn: SeqNum,
+    /// Capacity of backing storage
+    storage_cap: usize,
+    /// Cached prefix, not put into buffer until suffix is successfully read
+    stored_prefix: Option<E>,
+}
+
+impl<E, S> RaceReader<E, S>
+where
+    E: Copy + PartialEq + Entry,
+    S: Snapper<E>,
+{
+    /// Create new RaceReader
+    pub fn new(snapper: S, storage_cap: usize) -> RaceReader<E, S> {
+        RaceReader {
+            snapper,
+            read_seqn: 0.into(),
+            storage_cap,
+            stored_prefix: None,
+        }
+    }
+
+    /// Attempt to read all new entries in race buffer into given read buffer
+    /// Returns the number of entries missed before those that were read
+    pub fn read(&mut self, rbuf: &mut Vec<WholeEntry<E>>) -> Result<u64, S::Error> {
+        let pre_overwrite_seqn = self.snap_overwrite_seqn()?;
+        let pre_write_seqn = self.snap_write_seqn()?;
+
+        // If writer has overwritten unread entries between reads, store missed and correct read_seqn
+        let mut n_missed_before_read = num_missed(self.read_seqn, pre_overwrite_seqn);
+        self.read_seqn += n_missed_before_read;
+        // If any entries were missed and there is a stored prefix, then the entry after the prefix was missed.
+        // The prefix is dropped and added to the missed count.
+        // Note: the read sequence number is updated above because the prefix was included in the last read
+        if Into::<u64>::into(n_missed_before_read) > 0 && self.drop_prefix() {
+            n_missed_before_read += 1;
+        }
+
+        // Perform reads into snapshot buffer up to write sequence number
+        let first_read_seqn = self.read_seqn;
+        let mut buf_snapshot: Vec<E> = Vec::new();
+        while self.read_seqn != pre_write_seqn {
+            let storage_index = get_seqn_index(self.storage_cap, self.read_seqn, false);
+            buf_snapshot.push(self.snap_storage(storage_index)?);
+            self.read_seqn += 1;
+        }
+
+        // Calculate number of entries in snapshot buffer that may have been overwritten
+        let post_overwrite_seqn = self.snap_overwrite_seqn()?;
+        let n_overwritten_in_snap = min(
+            num_missed(first_read_seqn, post_overwrite_seqn),
+            (buf_snapshot.len() as u64).into(),
+        );
+        // If any entries were missed and there is a stored prefix, then the entry after the prefix was missed.
+        // The prefix is dropped and added to the missed count (not overwritten during snap)
+        if Into::<u64>::into(n_overwritten_in_snap) > 0u64 && self.drop_prefix() {
```

Comment:
FWIW `u64::from(n_overwritten_in_snap)` is a more compact and equivalent way to write the conversion.